### DISCUSSION TITLE
allow optimized compare to receive a CTE

### DISCUSSION THREAD

**Discussion by heckad** [sqlalchemy/sqlalchemy#9946] [Blue Sky]
I have this query
```
receiver_user = session.query(User).filter_by(token=user_receiver_token)

taken_collectible =session.query(TakenCollectible)\
    .filter_by(user=receiver_user, collectible_id=collectible_id) \
    .one_or_none()
```

What I expect:
I expect that sqlalchemy compile this request, but I got error
```
    def _optimized_compare(
        self,
        state,
        value_is_parent=False,
        adapt_source=None,
        alias_secondary=True,
    ):
        if state is not None:
>           state = attributes.instance_state(state)
E           AttributeError: 'BaseQuery' object has no attribute '_sa_instance_state'
```
Is there any way to use Query in another query like in Django, or need rewrite query on explicit using join?

**Comment by zzzeek** [comment]
I believe you need to convert receiver_user to a scalar subquery first:

    receiver_user = session.query(User).filter_by(token=user_receiver_token).as_scalar()

the relationship of objects here will become clearer in SQLAlchemy 2.0.

**Comment by heckad** [comment]
I got error new error ```AttributeError: 'Select' object has no attribute '_sa_instance_state'```

**Comment by zzzeek** [comment]
oh, OK this is a many-to-one relationship, and those do not support direct comparison to a SQL query right now.  Interesting idea though I've never seen anyone try that before.   I'm not sure how to easily support that as it would need to extract the correct columns from the SELECT query and that would be kind of complicated.

your two options are:

1. just load the object:

  receiver_user = session.query(A).filter_by(data="some data").one()


2. use the foreign key column in your comparison, not the relationship:

```
receiver_user = session.query(A.id).filter_by(data="some data").as_scalar()

taken_collectible = (
    session.query(B)
    .filter_by(a_id =receiver_user, data="some other data")
    .one_or_none()
)

```

**Comment by heckad** [comment]
I just wanted to avoid two queries. How to get something like
```
select * from taken_collectible where user_id = (select id from users where token='123') and collectible_id=1;
```
?

> it would need to extract the correct columns from the SELECT query and that would be kind of complicated.

Where is information about the field that the relationship is based on stored? I would like to implement this feature.

Thank you for your quick answers!

**Comment by zzzeek** [comment]
the second form.  compare to foreign key column.

**Comment by zzzeek** [comment]
just tried a few things and unfortunately there's no way to generalize this for composite primary keys at least.   it could work for a single-col key, but if we're trying to get a tuple, even PostgreSQL seems to reject this syntax:
```

(psycopg2.errors.SyntaxError) subquery has too few columns
LINE 3: WHERE (b.a_id, b.a_id2) = (SELECT (a.id, a.id2) AS anon_1 
                                ^

[SQL: SELECT b.id AS b_id, b.a_id AS b_a_id, b.a_id2 AS b_a_id2, b.data AS b_data 
FROM b 
WHERE (b.a_id, b.a_id2) = (SELECT (a.id, a.id2) AS anon_1 
FROM a 
WHERE a.data = %(data_1)s) AND b.data = %(data_2)s]

```

for now _optimized_compare can have a nicer message

**Comment by zzzeek** [comment]
oh also you can use has(), duh, sorry:

```
taken_collectible = (
    session.query(B)
    .filter(B.user.has(data="some data"))
    .filter_by(data="some other data")
    .one_or_none()
)

```

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Raise informative error when non-object m2o comparison used** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1903

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_3** branch:

**Raise informative error when non-object m2o comparison used** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1904

**Comment by heckad** [comment]
You can use cte for tuple of keys. Example

```
with subquery as (
    select a_id, b_id
    from table_1
    where ...
)
select * 
from table_2
where a_id = subquery.a_id and b_id = subquery.b_id;
```
or
```
select * 
from table_2, (
    select a_id, b_id
    from table_1
    where ...
) as subquery 
where a_id = subquery.a_id and b_id = subquery.b_id;

**Comment by heckad** [comment]
@zzzeek, I proposed the solution for composite primary keys. Maybe not close this issue?

**Comment by zzzeek** [comment]
CTEs are not supported by prominent databases such as MySQL (EDIT: doh I'm wrong here too, MySQL 8 has them) and SQLite (EDIT: I''m wrong, sqlite has CTEs now) so I would rather not have to support a complicated ORM feature such as this one.    the ratio of complexity to usefulness is too high and users can instead use CTEs directly if they'd like to emit this SQL.

**Comment by zzzeek** [comment]
as an example, the current EXISTS operators like has() and any() took years to get right in every possible situation; that they work with inheritance, with subqueries, with association proxies, on and on.    SQLAlchemy makes any SQL possible but it's not always easy or worth it to make complex SQL pop out of a small operator automatically.

**Comment by heckad** [comment]
> CTEs are not supported by prominent databases such as MySQL and SQLite


This
```
select * 
from table_2, (
    select a_id, b_id
    from table_1
    where ...
) as subquery 
where a_id = subquery.a_id and b_id = subquery.b_id;
```
works fine in sqlite and mysql

**Comment by zzzeek** [comment]
a key aspect of the EXISTS operators is that the SQL they emit is entirely self-contained within the column expression.   That's the only reason they are possible, really.  there are no ORM query comparison operators that ever intend to modify the FROM clause or other clauses of the SELECT as that is extremely difficult if even possible to make work in all situations.

**Comment by zzzeek** [comment]
OK I just thought of a way the CTE approach here may be acceptable.  Pass in a CTE() directly, that way the optimized compare does not have to guess:
```

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)
    bs = relationship("B", backref="a")


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)


e = create_engine("sqlite://", echo="debug")
Base.metadata.create_all(e)

s = Session(e)

a1, a2, a3 = A(), A(), A()
b1 = B()
b1.a = a3

s.add_all([a1, a2, a3, b1])
s.commit()

a_query = s.query(A).filter_by(id=a3.id).cte()

q = s.query(B).filter_by(a=a_query)

assert q.one() is b1


```

output:

```
(SELECT a.id AS id, a.data AS data 
FROM a 
WHERE a.id = ?)
 SELECT b.id AS b_id, b.a_id AS b_a_id, b.data AS b_data 
FROM b, anon_1 
WHERE b.a_id = anon_1.id

```


here's a totally not ready for primetime patch against master that does the above:


```
diff --git a/lib/sqlalchemy/orm/relationships.py b/lib/sqlalchemy/orm/relationships.py
index 77e0930c6..0991c6263 100644
--- a/lib/sqlalchemy/orm/relationships.py
+++ b/lib/sqlalchemy/orm/relationships.py
@@ -1620,6 +1620,7 @@ class RelationshipProperty(StrategizedProperty):
         adapt_source=None,
         alias_secondary=True,
     ):
+
         if state is not None:
             try:
                 state = inspect(state)
@@ -1627,13 +1628,21 @@ class RelationshipProperty(StrategizedProperty):
                 state = None
 
             if state is None or not getattr(state, "is_instance", False):
-                raise sa_exc.ArgumentError(
-                    "Mapped instance expected for relationship "
-                    "comparison to object.   Classes, queries and other "
-                    "SQL elements are not accepted in this context; for "
-                    "comparison with a subquery, "
-                    "use %s.has(**criteria)." % self
-                )
+                from sqlalchemy.sql.expression import CTE, and_
+                if not isinstance(state, CTE):
+                    raise sa_exc.ArgumentError(
+                        "Mapped instance expected for relationship "
+                        "comparison to object.   Classes, queries and other "
+                        "SQL elements are not accepted in this context; for "
+                        "comparison with a subquery, "
+                        "use %s.has(**criteria)." % self
+                    )
+                else:
+                    return and_(
+                        local == state.corresponding_column(remote)
+                        for local, remote in self.local_remote_pairs
+                    )
+
         reverse_direction = not value_is_parent
 
         if state is None:

```

**Comment by zzzeek** [comment]
the general idea is simple but patching this correctly has a lot to take into account:

1. secondary relationships

2. clause adaption

3. tests: including tablea->tableb, self-referential use cases, joined table inheritance use cases, with_polymorphic use cases.   These would probably need to be in test/orm/test_relationships.py and test/orm/inheritance/test_relationships.py

**Comment by zzzeek** [comment]
also SQLite apparently has CTEs now since I ran that test on sqlite.    geez so does MySQL 8.  Well there you go.

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**Expand CTE testing and begin to make this more prominent.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1953

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**Expand CTE testing and begin to make this more prominent.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1954

**Comment by Ghost93** [comment]
Hey guys,

Just confirming that on PostgreSQL, the following syntax works when querying directly the database and when using SQLAlchemy it does not:
```WHERE (col1,col2) = (val1,val2)```

Exception is as follows:
```
(psycopg2.errors.SyntaxError) subquery has too few columns
LINE 5: ...                        AND ((col1, col2)) IN
                                                ^
 [SQL: 'MY_SQL_HERE '] [parameters: {my_dict_of_params_here}] 
(Background on this error at: http://sqlalche.me/e/f405)
```

Any updates on this?

**Comment by zzzeek** [comment]
hey there -

on a github issue, "updates" by definition would be comments here, so if you see nothing, that means "no" :)      there's no resources to attend to this issue right now and this is also low priority.   tagged with "motivated volunteers requested" indicates this.

**Comment by zzzeek** [comment]
im not sure how your code fragment relates to the issue at hand here.  to compare tuples on PostgreSQL you must use the tuple() API: https://docs.sqlalchemy.org/en/13/core/sqlelement.html?highlight=tuple_#sqlalchemy.sql.expression.tuple_

**Comment by Ghost93** [comment]
@zzzeek ok, thanks for the info.
I saw the `tuple()` on the API, problem I'm referring to is when direct raw sql is used, i.e. 
```
SELECT id
FROM table1 t
WHERE (t.col1,t.col2)=('a','b')
```

**Comment by zzzeek** [comment]
do you need to open a "question" issue for this?  im not sure how this is relevant to the topic here.   open a new issue with what you're having problems with, thanks.

**Comment by Ghost93** [comment]
I wrote here because it's the same error as you had on April - [link to the specific relevant comment](https://github.com/sqlalchemy/sqlalchemy/issues/5269#issuecomment-616222309)

**Comment by zzzeek** [comment]
this is a very esoteric use case, one can certainly use a CTE now and run it in a query to compare against many-to-one FK columns; the automatic conversion feature here would be a little bit magical and people would generally come across it by accident, a little tiny and extremely specific trick that happens and I think things are more straightforward not having little easter eggs in the code.
